### ISSUE TITLE
Upload speed in Windows may be diminished

### ISSUE THREAD

**Issue by jay** [curl/curl#2224]
### I did this

In Windows 7 I ran an upload speed test using an example called [PostUpload](https://gist.github.com/jay/daac01682f17c2dc213e) that I wrote for another issue and uses libcurl. When I modified libcurl to set the socket send buffer size to 64k the upload was twice as fast. 

By default in Vista or later libcurl does not set the socket send buffer (SO_SNDBUF).
~~~
Transfer rate: 1463 KB/sec (52428883 bytes in 35 seconds)
~~~

If the socket send buffer is set to 64k the upload speed on this Windows 7 computer is twice as fast.
~~~
Transfer rate: 3012 KB/sec (52428883 bytes in 17 seconds)
~~~
~~~diff
diff --git a/lib/connect.c b/lib/connect.c
index 3edb71e..d09ef01 100644
--- a/lib/connect.c
+++ b/lib/connect.c
@@ -943,7 +943,8 @@ static void nosigpipe(struct connectdata *conn,
 
 void Curl_sndbufset(curl_socket_t sockfd)
 {
-  int val = CURL_MAX_WRITE_SIZE + 32;
+  int val = 64*1024;
+/*
   int curval = 0;
   int curlen = sizeof(curval);
 
@@ -963,7 +964,7 @@ void Curl_sndbufset(curl_socket_t sockfd)
   if(getsockopt(sockfd, SOL_SOCKET, SO_SNDBUF, (char *)&curval, &curlen) == 0)
     if(curval > val)
       return;
-
+*/
   setsockopt(sockfd, SOL_SOCKET, SO_SNDBUF, (const char *)&val, sizeof(val));
 }
 #endif
~~~

[Upload speed comparison #2224.zip](https://github.com/curl/curl/files/1613871/Upload.speed.comparison.2224.zip)
is the packet capture of both uploads. I don't have an automated script to parse the results but I can see that the common packet size is 16438 for both. In other words, they are both sending the same size packets most of the time from what I can tell.

This came out of an issue reported on the mailing list [upload speed is too slow in windows](https://curl.haxx.se/mail/lib-2018-01/0009.html) where a user claims to have a 6MB/sec upload speed in Chrome and yet libcurl they only get about 5% of that at 350k/sec. Setting the socket send buffer with an increased size gives them speed comparable to Chrome. It is extreme and I can't recall any other reports like it. I'm not sure why the difference that reporter is seeing is so drastic, I suspect something else may be going on as well.

### Other

[Design issues - Sending small data segments over TCP with Winsock](https://support.microsoft.com/en-us/help/214397/design-issues---sending-small-data-segments-over-tcp-with-winsock)

### I expected the following

The same results regardless of whether the socket send buffer is set.

### curl/libcurl version

master 908a9a6 2018-01-03

### operating system

Windows 7 Enterprise

**Comment by bagder** [comment]
So is 64K the most sensible on windows 7? Does 128K, 256KB or 512KB show any difference?

**Comment by jay** [comment]
> So is 64K the most sensible on windows 7? Does 128K, 256KB or 512KB show any difference?

I don't know if it's the most sensible. I think it depends on the upload bandwidth available. Since the time I wrote that example my upload bandwidth increased 2x and now to max out that bandwidth apparently I would have to set socket send at 64k. 128k and more no discernible difference since it's already maxed out. localhost I get the same speed regardless of whether or not I set the socket send buffer.

I'm not sure whether it's a good idea to increase the socket send buffer. Maybe this is the way Windows manages sharing the bandwidth? I have QoS turned off. Do you see any difference in Windows?

**Comment by MarcelRaad** [comment]
There was also a discussion a few years ago that led to the current logic of having the OS decide on Windows Vista and later in https://github.com/curl/curl/commit/43e045fc3e8430b2d2f3a1fc1354f3ecd54b09ab:
https://sourceforge.net/p/curl/bugs/1188/
That commit improved performance for me drastically back then.
There's probably no perfect solution.

**Comment by bagder** [comment]
Since the speed difference was so huge, I still think it is worth walking down this road a bit more before we give up because we don't know. At least we know that the increased buffer worked in one case. The question is perhaps now primarily if it can damage other use cases.

Can we produce a little test application and ask for win7 users to help us test-run from their systems and get some data on how this works from more places?

**Comment by wesdunn** [comment]
I’d be more than happy to help test on Win 7 systems. To be honest, I’m a fairly junior dev, but work on a product that utilizes and has utilized libcurl for quite a while. I’d really enjoy the opportunity to help out with an issue like this and do something to contribute to the project. I could help by either running a sample app and/or testing/benchmarking changes in test builds of our client.

**Comment by bagder** [comment]
I think we should start with writing up a test app that automatically tests a few different sizes, collects some stats and outputs the numbers on stdout or to a file. Then we can ask around for a bunch of win7 users to test run it and report back with their results and we would have data to base our decisions on. Hopefully those numbers will show us that some sizes are better than others and we could then do that by default on win7 in the future. (I think we should limit this to strictly be about win7 to keep things simple to start with, and then we can possibly expand it to other windows versions later)

You interested in putting together such a test @wesdunn ?

**Comment by jay** [comment]
The first thing you could do @wesdunn is run the PostUpload test or the curl tool equivalent, and then change it to 64k and comment out that section that returns on vista+ so that setsockopt is called (see the patch in the first post), and then run it again. When I do that I see a large difference, the upload speed is doubled.

make the 50MB of upload data
~~~
perl -e "print 'size_data=52428800&test_type=upload&svrPort=80'" > ulspeedtest
perl -e "print '&svrPort=80&start=9999999999999&data='" >> ulspeedtest
perl -e "print '0' x 52428800" >> ulspeedtest
~~~

run postupload or this curl tool equivalent:
~~~
curl --data-binary @ulspeedtest --write-out "Code %{http_code}, Upload speed: %{speed_upload} B/s\n" -o NUL http://ny2.testmy.net/uploader

rem you could run it this way which should show no difference:
curl -T ulspeedtest -X POST -H "Content-Type: application/x-www-form-urlencoded" --write-out "Code %{http_code}, Upload speed: %{speed_upload} B/s\n" -o NUL http://ny2.testmy.net/uploader
~~~

It's not accurate unless you get reply server code 200 like `Code 200, Upload speed: 1488399.000 B/s`. You can calculate Mbps from the B/s by dividing by 125000, for example `powershell 1488399.000/125000`. You can compare the results to each other and to the Mbps you see in the browser from the speedtest website. Proxies will skew the results.

**Comment by wesdunn** [comment]
@bagder I'd love the opportunity! I'm pretty slammed with my day job right now, so I may not have a lot of time to dedicate to this for a week or so. If this is an urgent issue, then maybe someone else should pick it up, I don't want to delay anything. 

I can definitely take a couple hours this week and go through what @jay recommended and report back with my results. Would that work to get things started?

@jay Thanks for the assist!

**Comment by djelinski** [comment]
There is no single optimal buffer size. The buffer size determines how much data the sender can put on the line, and the optimal value is equal to bandwidth-delay product (BDP).
Before Windows Vista SP1 it was the client's responsibility to make a guess and set this value. Starting with Vista SP1 we can let the OS determine optimal buffer size using one of the APIs documented here:
https://msdn.microsoft.com/en-us/library/windows/desktop/bb736549(v=vs.85).aspx
Also, starting with Windows 7 the system manages the send buffer size. There's only one little gotcha, the buffer management feature is silently disabled when socket is put into nonblocking mode. Curl does that on every socket, so won't be able to take advantage of this feature.

**Comment by jay** [comment]
> Also, starting with Windows 7 the system manages the send buffer size. There's only one little gotcha, the buffer management feature is silently disabled when socket is put into nonblocking mode.

AFAIK libcurl doesn't set SO_SNDBUF by default on Vista+ so we should get the system managed size. Where does it say that the management is disabled in non-blocking mode? I was unaware of that...

**Comment by djelinski** [comment]
> AFAIK libcurl doesn't set SO_SNDBUF by default on Vista+

You're right, it doesn't.
> we should get the system managed size

That's my understanding of the documentation as well. But we don't, at least on Windows Server 2008R2. In [my mail](https://curl.haxx.se/mail/lib-2018-07/0080.html) I provided sample code that I used to verify that blocking sockets are and nonblocking sockets are not managed. 
> Where does it say that the management is disabled in non-blocking mode?

It doesn't, so the behavior might change between Windows versions & patches.
